PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/TelemetryClientProvider.java

```diff
@@ -0,0 +1,330 @@
+     * @return A PsiElementVisitor object that visits the method calls in the code.
+     */
+    @NotNull
+    @Override
+    public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
+
+        // Reset the methodCounts map for each new visitor
+        TelemetryClientProviderVisitor.methodCounts.clear();
+        TelemetryClientProviderVisitor visitor = new TelemetryClientProviderVisitor(holder, isOnTheFly);
+
+        // Set the visitor instance in TelemetryToggleAction
+        TelemetryToggleAction.setTelemetryService(visitor);
+
+        return visitor;
+    }
+
+    /**
+     * This class is a visitor that visits the method calls in the code and tracks the method calls.
+     */
+    static class TelemetryClientProviderVisitor extends JavaElementVisitor {
+
+        private final ProblemsHolder holder;
+        private static boolean running = false; // Flag to indicate if the telemetry service is running
+        private static ScheduledExecutorService executorService; // Executor service to schedule telemetry data sending
+
+        // Create a TelemetryClient object
+        // not final because the test involves Injecting the mock telemetry client to telemetryClient
+        // Package-private to allow access from tests in the same package
+        static TelemetryClient telemetryClient = getTelemetryClient();
+        // Create a map to store the method counts
+        static Map<String, Map<String, Integer>> methodCounts = new HashMap<>();
+
+        // Create a Project object
+        private static Project project;
+
+        private static final List<String> AZURE_METHOD_PREFIXES = Arrays.asList("upsert", "set", "create", "update", "replace", "delete", "add", "get", "list", "upload");
+
+        // Create a logger object
+        private static final Logger LOGGER = Logger.getLogger(TelemetryClientProvider.class.getName());
+
+        // Create a RuleConfig object
+        private static final RuleConfig ruleConfig;
+
+
+        static {
+            final String ruleName = "TelemetryClientProvider";
+            RuleConfigLoader centralRuleConfigLoader = RuleConfigLoader.getInstance();
+
+            // Get the RuleConfig object for the rule
+            // There is no rule associated with this inspection, so the ruleConfig is set to an empty RuleConfig object
+            ruleConfig = centralRuleConfigLoader.getRuleConfig(ruleName);
+        }
+
+        /**
+         * This constructor is used to create a visitor for the inspection
+         * It initializes the holder and isOnTheFly fields.
+         *
+         * @param holder     The ProblemsHolder object that holds the problems found in the code.
+         * @param isOnTheFly A boolean that indicates if the inspection is running on the fly. - This is not used in this implementation.
+         */
+        public TelemetryClientProviderVisitor(ProblemsHolder holder, boolean isOnTheFly) {
```

Comment:
does this need to be public?

Reply:
oh package-private is good too I changed it